Create the concept of embeddableHandlers

stacey-gammon opened the pull request:
Register embeddable handlers to push rendering logic outside of dashboard.

As dashboard attempts to render each panel, it will use panel type to grab the corresponding embeddable handler.  The actual rendering logic is inside of the embeddableHandler which gets passed a dom node, along with some actions.

Stacey-Gammon commented:
jenkins, test this

Stacey-Gammon commented:
@thomasneirynck @kjbekkelund, do you guys think you will have a chance to review this over the next week or so?  It's not a big rush but would help move things along if I could get it checked in.

kjbekkelund commented on src/core_plugins/kibana/public/dashboard/panel/panel.js:
```diff
@@ -72,116 +66,69 @@ uiModules
-        }
+        const type = $scope.panel.type;
+        const id = $scope.panel.id;
+        const service = _.find(services, { type: type });
```
nitpick (and I know it's just copy/pasted from the old code): If we're still on the "use less lodash", this can be `services.find(service => service.type === type)`.

Stacey-Gammon commented:
@spalger - do you have interest in giving this a look over?

Stacey-Gammon commented:
> I would consider creating an abstract class of an EmbeddableHandler that defines the shape of the type. All children (Visualize/Search/...) then subclass from it. The duck-typing of the EmbedablleHandler-type now works, but having the inheritance would help navigability and would be a form of documentation too.

@thomasneirynck  - I don't think we have any precedent for this in our code base currently.  I'm personally a fan of it, and I used interfaces with typescript at a prior company, but when I've done that type of thing here, in PRs past, it's generally been unfavorable viewed, and the comprise decided on was to use JSDoc to document the interface. The downside being that you need a common place to put the interface and there isn't a good common spot right now.  I suppose I could put it in the embeddable_handler_registery.  What do you think of that approach?

  cc @weltenwort since I think most of the conversations I had regarding this topic were with you, can you confirm that I'm stating that correctly and you prefer JSDoc interface definitions over abstract classes?  Perhaps we should bring this up as a new issue for team wide votes so we can have the decision written down and everyones input.

Stacey-Gammon commented:
@thomasneirynck 
>I'd also add, beyond the 'render', `getEditPath', ... two new methods:
a destroy method, that gets called for all when people navigate away from the dashboard.
a resize(), that instructs an embeddalbe to fit to the dimension of the new surrounding container.

Right now each embeddable handles destroying itself when that happens:
```
visualizationInstance.on('$destroy', function () {...}
```

If we wanted to extract it out into a new function in the handler, we'd have to keep track of all panels that were rendered, and map them to their scope objects, because a destroy on a handler will need to destroy every rendered panel.

Maybe what we need, in the long term, is to have and Embeddable and an EmbeddableHandler, and the Embeddable has a destroy method.  But as of right now I think that would just complicate the logic.

The same thing goes for the resize function  - you'd need a context of all the panels the handler rendered, which right now, it doesn't need to store.  The resizing all works naturally, so I'm not even sure when I would call it in dashboard.

Stacey-Gammon commented:
@thomasneirynck 
> For each EmbeddableHandler, consider making name and title constants in the module-scope. These are immutable properties, so they don't need to be instance-properties assigned in the cosntructor.

I got rid of title for now, it wasn't being used. I'll bring it back when it is.  As for name, I'm pretty sure I need to leave it as is for the registry handlers to work correctly:
`const embeddableHandler = embeddableHandlers.byName[$scope.panel.type];`

Stacey-Gammon commented:
@thomasneirynck - should finally be ready for another look.  I added a base class with comments.  I didn't add a destroy or resize method because they aren't needed now - not to say that there shouldn't be more refactoring done in the future that would require their creation.  It sounds like a good idea on the surface but for this iteration, I don't see them being necessary.

Should be all merged with the vis refactor. I don't plan on submitting until 6.1 branch is created so it won't get checked in last minute to 6.0.

@spalger - any chance you can give this a look too and let me know if the plugin stuff I added meshes well with the new platform?  I followed suit with some of the ui export changes we made last week.

I plan on expanding the concept of embeddable handlers and further ripping object specific knowledge out of dashboard, but feel this is a good step 1.

spalger commented on src/core_plugins/kibana/public/dashboard/panel/panel.js:
```diff
@@ -89,103 +74,68 @@ uiModules
-        $scope.savedObj = savedObjectInfo.savedObj;
-        $scope.editUrl = savedObjectInfo.editUrl;
+      const savePanelState = (panel) => {
+        $scope.panel = Object.assign($scope.panel, panel);
```
Object.assign returns its first argument, so assigning here isn't necessary, but I think it makes sense to merge instead of mutate:

```js
const updatePanelState = (changes) => {
  $scope.panel = {
    ...$scope.panel,
    ...panel
  }
  $scope.saveState();
}
```

thomasneirynck commented on src/ui/public/embeddable/embeddable_handler.js:
```diff
@@ -0,0 +1,35 @@
+ */
+export class EmbeddableHandler {
+  /**
+   * @param {string} panelId - the id of the panel to grad the title for.
```
what's to grad?

thomasneirynck commented on src/ui/public/embeddable/embeddable_handler.js:
```diff
@@ -0,0 +1,35 @@
+   * from panel.id because you can have the same object rendered multiple times in two different panels.
+   * @param {Object} container - Contains functions to communicate with the container.
+   */
+  render(/* domNode, panel, container */) {
```
consider requiring `render` to be async. It's an easy compilation-time flag to require users to (optionally) return a Promise. similar for getTitle, getEditPath

thomasneirynck commented on src/ui/public/embeddable/embeddable_handler.js:
```diff
@@ -0,0 +1,35 @@
+
+  /**
+   * @param {Element} domNode - the dom node to mount the rendered embeddable on
+   * @param {Object} panel - a panel object which container information about the panel. Can also be modified to
```
`@param {PanelState} panelState` instead?

thomasneirynck commented on src/core_plugins/kibana/public/dashboard/panel/panel.js:
```diff
@@ -89,103 +74,72 @@ uiModules
-        // create child ui state from the savedObj
-        const uiState = $scope.savedObj.uiStateJSON ? JSON.parse($scope.savedObj.uiStateJSON) : {};
-        $scope.uiState = $scope.createChildUiState(getPersistedStateId($scope.panel), uiState);
+      const containerAPI = {
```
consider extracting to separate module and documenting with JSdoc.

Stacey-Gammon commented:
@spalger -- any more feedback on your end?

spalger commented on src/ui/public/embeddable/embeddable_handlers_registry.js:
```diff
@@ -0,0 +1,9 @@
+import { uiRegistry } from 'ui/registry/_registry';
```
Can we put this in `ui/registry` with the rest of the registries? I'm not a huge fan of the decision to go that route but we should probably stay consistent.

Stacey-Gammon replied:
We are already starting to deviate elsewhere - savedObjectsRegistery.  Though yes most are in the ui/registry folder.  If you feel strongly about it I'll move it over to ui/registry, though I'd rather not. :)

spalger commented on src/ui/public/embeddable/container_api.js:
```diff
@@ -0,0 +1,58 @@
+   * @param operator
+   * @param index
+   */
+  onFilter(/*field, value, operator, index */) {
```
This should probably be `addFilter()` right?

spalger commented on src/ui/public/embeddable/container_api.js:
```diff
@@ -0,0 +1,58 @@
+   * @param uiState {Object} - the uiState for the child.
+   * @returns {Object}
+   */
+  createChildUiState(/* path, uiState */) {
```
Can we call this `initalState` or something that disambiguates it from the `PersistedState` instances we usually call `uiState`? If it was a `PersistedState` then we would just be able to call `createChild()` on it.

spalger commented on src/ui/public/embeddable/container_api.js:
```diff
@@ -0,0 +1,58 @@
+/**
+ * The ContainerAPI is an interface for embeddable objects to interact with the container they are embedded within.
+ */
+export class ContainerAPI {
```
👍

spalger commented on src/core_plugins/kibana/public/discover/embeddable/search_embeddable_handler.js:
```diff
@@ -0,0 +1,77 @@
+    return savedObject.title;
+  }
+
+  async render(domNode, panel, container) {
```
Unfortunately there are a lot of unexpected and hard-to-debug issues that pop-up when native promises or async/await with Angular. I hate to say it but please convert any `async/await` or native promise usage that is interacting with Angular to use Angular promises.

Stacey-Gammon replied:
so saaaaaaad :(

spalger commented on src/ui/public/embeddable/embeddable_handler.js:
```diff
@@ -0,0 +1,36 @@
+   * from panel.id because you can have the same object rendered multiple times in two different panels.
+   * @param {Promise.<void>} A promise that resolves when the object is finished rendering.
+   */
+  async render(/* domNode, panel, container */) {
```
Looks like the params here are now `domNode, panel, containerApi`? jsdocs need updates.

spalger commented on src/core_plugins/kibana/public/discover/embeddable/search_embeddable_handler.js:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  async getEditPath(panelId) {
+    return await this.searchLoader.urlFor(panelId);
```
If an embeddable had controls for editing itself it would need to be able to tell consumers that the title/path/etc. have changed, any idea how that would work?

I assume we couldn't just replace these methods with something like `ContainerAPI#setEditPath()` and `ContainerAPI#setTitle()` because we want to show the edit path and title for a panel without rendering it... Maybe the container API could expose a method to invalidate the editPath and titles that are cached by containers?

Stacey-Gammon replied:
Yea, that's a good point. Not sure how I feel about adding that functionality now though when it's not needed by anything we build.  Yes, we want this to be a flexible embeddable system for others to easily write panels in, but I think during this initial phase, it's okay to limit what the container api exposes.   We should probably add some kind of forceRefresh functionality that forces a rebuild of the whole panel at some point. Just not sure I want to go down this path without an easy way to test it works.  Thoughts?

spalger replied:
Yeah, without a use case we probably shouldn't design a solution, I'm fine with punting for now. We can figure out how to make it work when we need it (hint: I'm thinking observables)

spalger commented on src/core_plugins/kibana/public/dashboard/dashboard_container_api.js:
```diff
@@ -0,0 +1,38 @@
+    return this.dashboardState.getIsViewMode();
+  }
+
+  getInitialState(path, uiState) {
```
Haha, I wasn't super clear, I was asking for `createChildUiState(path, initialState)`

This method returns an instance of `PersistedState`, which we usually refer to as `uiState`. This method also takes an argument which it calls `uiState`, but it's actually just a plain object that is used as the initial state for the `uiState` that is returned.

Stacey-Gammon replied:
hah, okay, yea, this function is carried over from the past and I never understood much of it.  Will switch to your intended version. :)

spalger commented on src/ui/public/embeddable/embeddable_handler.js:
```diff
@@ -0,0 +1,33 @@
+   * @return {Promise.<string>} a promise that resolves with the path that dictates where the user will be navigated to
+   * when they click the edit icon.
+   */
+  async getEditPath(/* panelId */) {
```
No need for `async` (or `Promise.reject` below)

spalger commented:
Looks like the spy panel isn't visible on the dashboard anymore:

master:
![2017-08-03 14 47 41](https://user-images.githubusercontent.com/1329312/28945346-f088d1a8-785a-11e7-8a7f-f733e38cd2cc.gif)


pr:
![2017-08-03 14 48 51](https://user-images.githubusercontent.com/1329312/28945342-ef373880-785a-11e7-9f43-de44faf987cd.gif)

spalger commented:
Looks like spy panel shows up when visualizations are added, sounds like a two-way binding is being broken

Stacey-Gammon commented:
Nice find.  I see where it's coming from.  Strange though because I have a test that verifies the spyPaneToggle button exists, and works, and that should have caught it.

Will investigate.

Stacey-Gammon commented:
Ahhh, no it's bc it's checking view mode when it's supposed to be checking full screen mode.   That's why the tests are passing bc it's in edit mode when they run.

Stacey-Gammon commented:
Added a test that would have caught it and fixed up the logic.  The binding part is still an issue, but it's an issue in master as well and I filed https://github.com/elastic/kibana/issues/13340 to keep track.  It's just much less obvious of a bug when the logic is correct because it only hides the spy pane in embed mode and you aren't supposed to leave embed mode without a hard refresh.
